test/fuzz/transformation_add_dead_block_test.cpp

```diff
@@ -142,6 +142,101 @@ TEST(TransformationAddDeadBlockTest, BasicTest) {
   ASSERT_TRUE(IsEqual(env, variant_shader, context.get()));
 }
 
+TEST(TransformationAddDeadBlockTest, Applicable) {
+  std::string reference_shader = R"(
+               OpCapability Shader
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %6 "main"
+               OpExecutionMode %6 OriginUpperLeft
+
+; Types
+          %2 = OpTypeBool
+          %3 = OpTypeVoid
+          %4 = OpTypeFunction %3
+
+; Constants
+          %5 = OpConstantTrue %2
+
+; main function
+          %6 = OpFunction %3 None %4
+          %7 = OpLabel
+               OpSelectionMerge %11 None
+               OpBranchConditional %5 %8 %9
+          %8 = OpLabel
+               OpBranch %10
+          %9 = OpLabel
+               OpBranch %10
+         %10 = OpLabel
+               OpBranch %11
+         %11 = OpLabel
+               OpBranch %13
+         %12 = OpLabel
+               OpBranch %13
+         %13 = OpLabel
+               OpReturn
+               OpFunctionEnd
+  )";
+  const auto env = SPV_ENV_UNIVERSAL_1_4;
+  const auto consumer = nullptr;
+  const auto context =
+      BuildModule(env, consumer, reference_shader, kFuzzAssembleOption);
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+  auto transformation = TransformationAddDeadBlock(14, 11, false);
+  ASSERT_FALSE(
+      transformation.IsApplicable(context.get(), transformation_context));
+
+  ApplyAndCheckFreshIds(transformation, context.get(), &transformation_context);
```

Comment:
A key part of the design of spirv-fuzz is that transformations should only be applied if `IsApplicable` holds. The `Apply` method (which is called by `ApplyAndCheckFreshIds`) assumes that `IsApplicable` holds.

Here, you are applying a transformation for which `IsApplicable` returns *false*, and that's wrong.

Do you need to apply the transformation? If the missing coverage was in branches of `IsApplicable` that return false then it would be fine to have a test that merely creates a bad transformation and asserts that it is not applicable.

Reply:
Yes, @afd, you are right, I want to apply the transformation, I've followed your suggestion, and I think that's it: 
```CPP
   
    TEST(TransformationAddDeadBlockTest, NotApplicable) {
  std::string reference_shader = R"(
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %6 "main"
               OpExecutionMode %6 OriginUpperLeft

; Types
          %2 = OpTypeBool
          %3 = OpTypeVoid
          %4 = OpTypeFunction %3

; Constants
          %5 = OpConstantTrue %2

; main function
          %6 = OpFunction %3 None %4
          %7 = OpLabel
               OpSelectionMerge %11 None
               OpBranchConditional %5 %8 %9
          %8 = OpLabel
               OpBranch %10
          %9 = OpLabel
               OpBranch %10
         %10 = OpLabel
               OpBranch %11
         %11 = OpLabel
               OpBranch %13
         %12 = OpLabel
               OpBranch %13
         %13 = OpLabel
               OpReturn
               OpFunctionEnd
  )";
  const auto env = SPV_ENV_UNIVERSAL_1_4;
  const auto consumer = nullptr;
  const auto context =
      BuildModule(env, consumer, reference_shader, kFuzzAssembleOption);
  spvtools::ValidatorOptions validator_options;
  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
                                               kConsoleMessageConsumer));
  TransformationContext transformation_context(
      MakeUnique<FactManager>(context.get()), validator_options);
  auto transformation = TransformationAddDeadBlock(14, 11, false);

  // Constant with same value of condition value not present.
  ASSERT_FALSE(
      transformation.IsApplicable(context.get(), transformation_context));

  ApplyAndCheckFreshIds(transformation, context.get(), &transformation_context);

  std::string variant_shader = R"(
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %6 "main"
               OpExecutionMode %6 OriginUpperLeft

; Types
          %2 = OpTypeBool
          %3 = OpTypeVoid
          %4 = OpTypeFunction %3

; Constants
          %5 = OpConstantTrue %2

; main function
          %6 = OpFunction %3 None %4
          %7 = OpLabel
               OpSelectionMerge %11 None
               OpBranchConditional %5 %8 %9
          %8 = OpLabel
               OpBranch %10
          %9 = OpLabel
               OpBranch %10
         %10 = OpLabel
               OpBranch %11
         %11 = OpLabel
               OpSelectionMerge %13 None
               OpBranchConditional %5 %13 %14
         %14 = OpLabel
               OpBranch %13
         %12 = OpLabel
               OpBranch %13
         %13 = OpLabel
               OpReturn
               OpFunctionEnd
  )";

  ASSERT_FALSE(fuzzerutil::IsValidAndWellFormed(
      context.get(), validator_options, kConsoleMessageConsumer));
  ASSERT_FALSE(IsEqual(env, variant_shader, context.get()));
}

```
If I'm right! I will push it.

Reply:
In the above comment, I still see:

```c++
  ASSERT_FALSE(
      transformation.IsApplicable(context.get(), transformation_context));

  ApplyAndCheckFreshIds(transformation, context.get(), &transformation_context);
```

In other words, IsApplicable will return false, and then you will call Apply. Thus, your test is doing something invalid (according to our rules about how transformations must be used). Based on the test name ("NotApplicable"), it looks like you want to test that IsApplicable returns false; you have done that. You must not apply that transformation. If you want to get coverage of something in `Apply` then write another test. Or, if you think the additional coverage should be part of this test, then create another transformation such that IsApplicable will return true so that you can then apply the transformation.

Reply:
OK, Thanks @paulthomson, I will adjust it.

Reply:
I've understood what is the problem I made and fix the test. Could you please review this @afd, @paulthomson?
```CPP

      TEST(TransformationAddDeadBlockTest, Applicable) {
        std::string reference_shader = R"(
                     OpCapability Shader
                %1 = OpExtInstImport "GLSL.std.450"
                     OpMemoryModel Logical GLSL450
                     OpEntryPoint Fragment %6 "main"
                     OpExecutionMode %6 OriginUpperLeft
      
      ; Types
                %2 = OpTypeBool
                %3 = OpTypeVoid
                %4 = OpTypeFunction %3
      
      ; Constants
                %5 = OpConstantFalse %2
      
      ; main function
                %6 = OpFunction %3 None %4
                %7 = OpLabel
                     OpSelectionMerge %11 None
                     OpBranchConditional %5 %8 %9
                %8 = OpLabel
                     OpBranch %10
                %9 = OpLabel
                     OpBranch %10
               %10 = OpLabel
                     OpBranch %11
               %11 = OpLabel
                     OpBranch %13
               %12 = OpLabel
                     OpBranch %13
               %13 = OpLabel
                     OpReturn
                     OpFunctionEnd
        )";
        const auto env = SPV_ENV_UNIVERSAL_1_4;
        const auto consumer = nullptr;
        const auto context =
            BuildModule(env, consumer, reference_shader, kFuzzAssembleOption);
        spvtools::ValidatorOptions validator_options;
        ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
                                                     kConsoleMessageConsumer));
        TransformationContext transformation_context(
            MakeUnique<FactManager>(context.get()), validator_options);
        auto transformation = TransformationAddDeadBlock(14, 11, false);
      
        ASSERT_TRUE(
            transformation.IsApplicable(context.get(), transformation_context));
      
        ApplyAndCheckFreshIds(transformation, context.get(), &transformation_context);
      
        std::string variant_shader = R"(
                     OpCapability Shader
                %1 = OpExtInstImport "GLSL.std.450"
                     OpMemoryModel Logical GLSL450
                     OpEntryPoint Fragment %6 "main"
                     OpExecutionMode %6 OriginUpperLeft
      
      ; Types
                %2 = OpTypeBool
                %3 = OpTypeVoid
                %4 = OpTypeFunction %3
      
      ; Constants
                %5 = OpConstantFalse %2
      
      ; main function
                %6 = OpFunction %3 None %4
                %7 = OpLabel
                     OpSelectionMerge %11 None
                     OpBranchConditional %5 %8 %9
                %8 = OpLabel
                     OpBranch %10
                %9 = OpLabel
                     OpBranch %10
               %10 = OpLabel
                     OpBranch %11
               %11 = OpLabel
                     OpSelectionMerge %13 None
                     OpBranchConditional %5 %14 %13
               %14 = OpLabel
                     OpBranch %13
               %12 = OpLabel
                     OpBranch %13
               %13 = OpLabel
                     OpReturn
                     OpFunctionEnd
        )";
      
        ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(
            context.get(), validator_options, kConsoleMessageConsumer));
        ASSERT_TRUE(IsEqual(env, variant_shader, context.get()));
      }

```

Reply:
Looks good. I guess you are trying to cover the the use of OpConstantFalse? If so, maybe make it clear from the test name and/or with a comment above where you create the transformation.

Reply:
The change looks good to me. Alternatively, you could've changed the `%5 = OpConstantFalse %2` to `OpConstantTrue` and specified `true` instead of `false` here `auto transformation = TransformationAddDeadBlock(14, 11, false);`. That would also make the transformation applicable.

Reply:
Yes, you are right @Vasniktel, but it will not handle the coverage case in the `apply` method in `transformation_add_dead_block.cpp`